### PR TITLE
Feature/components

### PR DESCRIPTION
**Pull request by chrislawlortech**
New components added with stories:
- ListHeader
- RackingSlice

Implementation of redux with use of redux-toolkit

### CODE REVIEW

File: app/components/common/RatingSlice.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+const RatingSlice = ({ title, redValue, amberValue, greenValue }) => {
+  return (
+    <View style={styles.container}>
+      <View style={styles.titleContainer}>
+        <RNText style={styles.title}>{title}</RNText>
+      </View>
+      <View style={styles.row}>
+        <View style={styles.columnHeader}>
+          <RNText style={styles.redTitle}>Red</RNText>
+        </View>
```
Comment (dantay0803) [inline app/components/common/RatingSlice.js]: You could look a splitting this into its own component. Would clean up this file by removing a few duplicate lines. For the style I would pass in the colour from the redTitle etc. But have a default that is aways applies to handle setting the font size.

File: app/components/common/RatingSlice.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      </View>
+      <View style={styles.breakerContainer}>
+        <View style={styles.horizontalLine} />
+        <View style={styles.horizontalLine} />
+        <View style={styles.horizontalLine} />
+      </View>
+      <View style={styles.row}>
+        <View style={styles.cell}>
+          <RNText style={styles.title}>{redValue}</RNText>
+        </View>
```
Comment (dantay0803) [inline app/components/common/RatingSlice.js]: Again you could also look at splitting this into its own component. RatingSliceCell, just helps reduce the duplicate code in this file.

File: app/components/common/ListHeader.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import React from 'react'
+import { StyleSheet, TouchableOpacity, Image } from 'react-native'
+import colors from '../../config/colors'
```
Comment (dantay0803) [inline app/components/common/ListHeader.js]: Little tip. The project has a package.json file included that will let you import using app/config/colors. 
Helps if you ever split move code into or out of folders. You don't need to change them all, just for future. Saved me some time before.

Comment (chrislawlortech) [reply app/components/common/ListHeader.js]: ah yeah, I've done that in a few of the newer files, good to know, can definitely use that on econform

File: package.json [hunk 4]
```diff
@@ -39,7 +41,9 @@
     "react-redux": "^7.2.2",
     "redux": "^4.0.5",
     "redux-persist": "^6.0.0",
-    "redux-saga": "^1.1.3"
+    "redux-saga": "^1.1.3",
```
Comment (MDHayes) [inline package.json]: Don't need this

File: package.json [hunk 5]
```diff
@@ -22,12 +22,14 @@
     }
   },
   "dependencies": {
+    "-": "0.0.1",
```
Comment (MDHayes) [inline package.json]: What is this?

File: package.json [hunk 6]
```diff
@@ -39,7 +41,9 @@
     "react-redux": "^7.2.2",
     "redux": "^4.0.5",
     "redux-persist": "^6.0.0",
-    "redux-saga": "^1.1.3"
+    "redux-saga": "^1.1.3",
+    "redux-thunk": "^2.3.0",
+    "save": "^2.4.0"
```
Comment (MDHayes) [inline package.json]: What is this? Where is it used?

File: app/store/store.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import systems from './reducers/rackings'
+
+const reducer = {
+  rackings,
+  systems,
+}
+
+const store = configureStore({
+  reducer,
+  middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
```
Comment (dantay0803) [inline app/store/store.js]: I'm guessing you are waiting until you start using thunks before adding it?

**Review by MDHayes — COMMENTED** [review]
`app/store/reducers/*` are these files what we want? Do they map to the data structure from the api? There is a curl request and light documentation in the pdf I sent on Monday morning.

`app/api/api.js` I think you should be calling this index.js? Make imports easier

File: app/components/common/ListHeader.js (hunk 3 above)
Comment (dantay0803) [reply app/components/common/ListHeader.js]: Yeah I just seen after I left that.

File: app/store/reducers/rackings.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      state.data = {}
+    },
+    fetchRackingsFulfilled: (state, action) => {
+      state.loading = false
+      state.data = action.payload.data
+      state.errorMessage = null
+    },
+    fetchRackingsRejected: (state) => {
+      state.loading = false
+      state.errorMessage = 'Error data list'
```
Comment (dantay0803) [inline app/store/reducers/rackings.js]: I would update the error message to be more detailed and user friendly. You can then add a selector to use this as an error message to the user.

File: app/store/reducers/systems.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      state.data = {}
+    },
+    fetchSystemsFulfilled: (state, action) => {
+      state.loading = false
+      state.data = action.payload.data
+      state.errorMessage = null
+    },
+    fetchSystemsRejected: (state) => {
+      state.loading = false
+      state.errorMessage = 'Error data list'
```
Comment (dantay0803) [inline app/store/reducers/systems.js]: Again update the error message to be more detailed and user friendly. Then add a selector to use this as an error message to the user.

File: app/store/store.js (hunk 7 above)
Comment (chrislawlortech) [reply app/store/store.js]: Yeah, going to get the api sauce set up beforehand

File: package.json (hunk 6 above)
Comment (chrislawlortech) [reply package.json]: I think I accidentally ran `npm install redux-thunk - save` instead of `--save` or something, removing

**Comment by MDHayes** [comment]
@chrislawlortech looking at this there's a disconnect between the api and the store. I think you need to add the full api client, a thunk for the request and then tidy up the reducers. You need one of fetchSystems or fetchRacking - its one api endpoint.

Keep at it, getting there.

File: app/components/common/RatingSlice.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    fontSize: 16,
+  },
+  amberTitle: {
+    color: colors.amber,
+    fontSize: 16,
+  },
+  greenTitle: {
+    color: colors.green,
+    fontSize: 16,
+  },
```
Comment (dantay0803) [inline app/components/common/RatingSlice.js:72]: Not a priority but with the RatingCell component you can remove these styles and reduce the code in this file more.

File: app/navigation/Authenticator.js [hunk 11]
```diff
@@ -29,42 +38,18 @@ const Authenticator = () => {
...
             options={{
               title: 'Econform',
-              headerStyle: {
-                backgroundColor: colors.darkGrey,
-              },
-              headerTintColor: '#fff',
-              headerTitleStyle: {
-                fontWeight: 'bold',
-              },
+              ...options,
```
Comment (dantay0803) [inline app/navigation/Authenticator.js:41]: With the screen options you can pass this directly into the Stack.Navigator component using its screenOptions prop. Again not overly important just a tip to save you some work next time.

```
screenOptions={{
        ...options,
      }}
```

File: app/screens/Dashboard.js [hunk 12]
```diff
@@ -1,24 +1,43 @@
...
+                  onPress={() =>
+                    navigation.navigate('Info', {
+                      systemId: system.id,
+                      rackingChildren: system.children.rackings,
+                      title: name,
+                    })
+                  }
+                />
+              )
+            })}
```
Comment (dantay0803) [inline app/screens/Dashboard.js]: You could use a FlatList here to render these items. https://reactnative.dev/docs/flatlist
It has a lot of handy props such as setting a component to render if no data is available and adding pull to refresh.

File: app/screens/Info.js [hunk 13]
```diff
@@ -1,17 +1,43 @@
...
+                subTitle={rack.fields.short_description}
+                onPress={() =>
+                  navigation.navigate('Detail', {
+                    id: rack.id,
+                    title: rack.fields.name,
+                  })
+                }
+              />
+            )
+          })}
```
Comment (dantay0803) [inline app/screens/Info.js]: Again here using a FlatList component will help with not needing to manually mapping over the data.

File: app/screens/Info.js [hunk 14]
```diff
@@ -1,17 +1,43 @@
...
 import ListItem from '../components/common/ListItem'
+import { useSelector } from 'react-redux'
 
-const Info = ({ navigation }) => {
+const Info = ({ navigation, route }) => {
+  const rackingData = useSelector((state) => state.rackings.data)
+  const { rackings, systems } = rackingData
+  const { systemId, rackingChildren } = route.params
+  const system = systems.find((syst) => {
+    return syst.id === systemId
```
Comment (dantay0803) [inline app/screens/Info.js]: You could create a selector to select items by id. Gives you the benefit of reusing the function easily across components without rewriting it.

File: app/screens/Dashboard.js [hunk 15]
```diff
@@ -1,24 +1,46 @@
...
+        iconName={item.fields.icon.name}
+        imageUrl={item.fields.header_image}
+        onPress={() =>
+          navigation.navigate('Info', {
+            systemId: item.id,
+            rackingChildren: item.children.rackings,
+            title: item.fields.name,
+          })
+        }
+      />
```
Comment (dantay0803) [inline app/screens/Dashboard.js:33]: Break your FlatList render items into their own component.It just helps reduce the amount of code in your screens and keeps it more readable.

File: app/screens/Dashboard.js [hunk 16]
```diff
@@ -1,24 +1,46 @@
...
+    )
+  }
   return (
     <SafeAreaView style={styles.container}>
       <ScrollView
         contentInsetAdjustmentBehavior="automatic"
         style={styles.scrollView}
       >
```
Comment (dantay0803) [inline app/screens/Dashboard.js:41]: With a FlatList you wont need the ScrollView as the list will auto scroll within the view.

File: app/screens/Info.js [hunk 17]
```diff
@@ -1,17 +1,49 @@
...
+      <ListItem
+        title={item.fields.name}
+        subTitle={item.fields.short_description}
+        onPress={() =>
+          navigation.navigate('Detail', {
+            id: item.id,
+            title: item.fields.name,
+          })
+        }
+      />
```
Comment (dantay0803) [inline app/screens/Info.js:31]: Again break this into it's own component.

File: app/screens/Info.js [hunk 18]
```diff
@@ -1,17 +1,49 @@
...
+  }
   return (
     <SafeAreaView style={styles.container}>
       <ScrollView
         contentInsetAdjustmentBehavior="automatic"
         style={styles.scrollView}
       >
         <View style={styles.container}>
-          <ListItem onPress={() => navigation.navigate('Detail')} />
+          <ListHeader images={system.fields.header_image} />
```
Comment (dantay0803) [inline app/screens/Info.js:41]: Again you don't need the SrollView if the FlatList is handling listing out the content for the screen.

Have a look at this https://reactnative.dev/docs/flatlist#listheadercomponent
That will allow you to add a header component to the FlatList and will handle scrolling the header and list content.

File: app/screens/Detail.js [hunk 19]
```diff
@@ -1,18 +1,69 @@
...
+          key={index}
+          title={name}
+          content={'test content'}
+          redValue={red && red.value}
+          amberValue={amber && amber.value}
+          greenValue={green && green.value}
+        />
+      )
+    })
+  }
```
Comment (dantay0803) [inline app/screens/Detail.js:45]: I would move this component and it's logic to it's own file.

**Pull request merged**